pyk/src/pyk/cterm/symbolic.py

```diff
@@ -94,6 +96,28 @@ def kast_to_kore(self, kinner: KInner) -> Pattern:
     def kore_to_kast(self, pattern: Pattern) -> KInner:
         return kore_to_kast(self._definition, pattern)
 
+    def minimize_constraints(self, constraints: tuple[KInner, ...], pc: KInner) -> tuple[KInner, ...]:
+        """Minimize given branching constraints with respect to a given path condition."""
+        # Determine intersection between all returned sets of branching constraints
+        flattened_default = tuple(flatten_label('#And', c) for c in constraints)
+        intersection = flattened_default[0]
+        for i in range(1, len(flattened_default)):
+            intersection = [c for c in intersection if c in flattened_default[i]]
+            # If intersection is empty, there is nothing to be done
+            if not intersection:
+                return constraints
+        # Check if non-empty intersection is entailed by the path condition
+        dummy_config = self._definition.empty_config(sort=GENERATED_TOP_CELL)
+        pc_cterm = CTerm(dummy_config, constraints=[pc])
+        intersection_cterm = CTerm(dummy_config, constraints=intersection)
+        implication_check = self.implies(pc_cterm, intersection_cterm, bind_universally=True)
+        # The intersection is not entailed, there is nothing to be done
+        if implication_check.csubst is None:
+            return constraints
+        # The intersection is entailed and can be filtered out of the branching constraints
+        else:
+            return tuple(mlAnd(c for c in cs if c not in intersection) for cs in flattened_default)
```

Comment:
Note that getting a substitution back does not necessarily mean the implication is true/valid, the validity is signaled by the flag `satisfiable` (which is a historic misnomer and should be called `valid`). An implication can be valid without any substitution required (think `Config  => Config`), or else a substitution may exist but the resulting path conditions are not fulfilled.

IDK if `pyk` compensates for that.

Reply:
I think in the `Config => Config` case the substitution will be empty, rather than non-existent. The other case, in which we get a substitution but the implication is not valid, I'm less sure of, perhaps @ehildenb can clarify. What I did was copy how target subsumption was checked.

Also, I think that we should definitely rename `satisfiable` to `valid`.

Reply:
Actually, that information is not propagated to the class carrying the information. All I have is:
```
class CTermImplies(NamedTuple):
    csubst: CSubst | None
    failing_cells: tuple[tuple[str, KInner], ...]
    remaining_implication: KInner | None
    logs: tuple[LogEntry, ...]
```